lib/ddtrace/buffer.rb

```diff
@@ -24,6 +24,21 @@ def push(item)
       end
     end
 
+    def concat(items)
+      @mutex.synchronize do
+        return if @closed
+
+        # Segment items into underflow and overflow
+        underflow, overflow = overflow_segments(items)
+
+        # Concatenate items do not exceed capacity.
+        add_all!(underflow) unless underflow.nil?
+
+        # Iteratively replace items, to ensure pseudo-random replacement.
+        overflow.each { |item| replace!(item) } unless overflow.nil?
```

Comment:
Maybe another way to do this is to not `replace!` each item, but to `concat` all items, including overflow, then randomly drop X number of items until its back to max size again.

Might have the negative consequences of resizing the item array (memory allocation cost), or not guaranteeing the last item pushed will be kept (it could be randomly dropped.)